include/flashinfer/sampling.cuh

```diff
@@ -339,6 +340,116 @@ __device__ __forceinline__ void DeviceSamplingFromProb(
   aggregate += aggregate_local;
 }
 
+template <typename DType, typename IdType>
+struct DataAndIndex {
+  DType data;
+  IdType index;
+
+  __device__ DataAndIndex operator+(const DataAndIndex& other) const {
+    if (data > other.data) {
+      return {data, index};
+    } else {
+      return {other.data, other.index};
+    }
+  }
+  __device__ DataAndIndex& operator+=(const DataAndIndex& other) {
+    if (data > other.data) {
+      return *this;
+    } else {
+      data = other.data;
+      index = other.index;
+      return *this;
+    }
+  }
+};
+
+template <typename DType, uint32_t VEC_SIZE>
+__device__ __forceinline__ vec_t<DType, VEC_SIZE> GenerateGumbelNoise(uint64_t philox_seed,
+                                                                      uint64_t philox_offset,
+                                                                      uint64_t subsequence) {
+  curandStatePhilox4_32_10_t state;
+  vec_t<float, VEC_SIZE> noise;
+  constexpr float kEPSILON = 1e-20f;
+  constexpr float kLOG2 = 0.6931471806f;
+  auto log_func = [](float x) { return log2f(x) * kLOG2; };
```

Comment:
You don have to do this explicitly, compiler will do it automatically.

What do I mean by using log2 is that:
```
Suppose c = log2(e)
-log(-log(x)) = -log2(-log2(x) * c) * c = (-log2(-log2(x)) + -log2(c)) * c = -log2(-log2(x)) * c - (c * log2(c))
```

The last term c * log2(c) is a constant term that you can safely ignore, so you can just use -log2(-log2(x)) * c.

After adding epsilon, it become  -log2(-log2(x + eps) + eps) * c